kommunicate/src/main/java/io/kommunicate/KmConversationHelper.java

```diff
@@ -667,43 +682,47 @@ public void onFailure(ChannelFeedApiResponse channelFeedApiResponse, Context con
-        if (conversationBuilder == null) {
-            throw new KmException(Utils.getString(conversationBuilder.getContext(), R.string.km_conversation_builder_cannot_be_null));
-        }
-
-        if (conversationBuilder.getAgentIds() == null || conversationBuilder.getAgentIds().isEmpty()) {
-            KmCallback callback = new KmCallback() {
-                @Override
-                public void onSuccess(Object message) {
-                    KmAppSettingModel.KmResponse agent = (KmAppSettingModel.KmResponse) message;
-                    if (agent != null) {
-                        List<String> agents = new ArrayList<>();
-                        agents.add(agent.getAgentId());
-                        conversationBuilder.setAgentIds(agents);
-                        try {
-                            final String clientChannelKey = !TextUtils.isEmpty(conversationBuilder.getClientConversationId()) ? conversationBuilder.getClientConversationId() : (conversationBuilder.isSingleConversation() ? getClientGroupId(conversationBuilder.getUserIds(), agents, conversationBuilder.getBotIds(), conversationBuilder.getContext()) : null);
-                            if (!TextUtils.isEmpty(clientChannelKey)) {
-                                conversationBuilder.setClientConversationId(clientChannelKey);
-                                startOrGetConversation(conversationBuilder, handler);
-                            } else {
-                                createConversation(conversationBuilder, handler);
-                            }
-                        } catch (KmException e) {
-                            e.printStackTrace();
+    private static KmCallback getCallbackWithAppSettingsToCreateConversation(final boolean useSingleThreadedSettingFromServer, final KmConversationBuilder conversationBuilder, final KmStartConversationHandler handler) {
+        return new KmCallback() {
+            @Override
+            public void onSuccess(Object message) {
+                KmAppSettingModel.KmResponse kmAppSettings = (KmAppSettingModel.KmResponse) message;
+                if (kmAppSettings != null) {
+                    List<String> agents = new ArrayList<>();
+                    agents.add(kmAppSettings.getAgentId());
+                    conversationBuilder.setAgentIds(agents);
+                    if (useSingleThreadedSettingFromServer) {
+                        conversationBuilder.setSingleConversation(kmAppSettings.getChatWidget().isSingleThreaded());
+                    }
+                    try {
+                        final String clientChannelKey = !TextUtils.isEmpty(conversationBuilder.getClientConversationId()) ? conversationBuilder.getClientConversationId() : (conversationBuilder.isSingleConversation() ? getClientGroupId(conversationBuilder.getUserIds(), agents, conversationBuilder.getBotIds(), conversationBuilder.getContext()) : null);
```

Comment:
The priority is given to clientConversationId over single conversation if it is passed in the builder. Put this check above. SO if the builder has client conversation Id , there is no need to make the app setting call. Its an unnecessary call. Also the app setting call is already done once to get the agents, We can merge these two calls and also process the theme on success. This will not only reduce the duplicate call but also change the theme without need to logout /login.